### PR TITLE
Implement forward-filter backward-sample poutine

### PR DESCRIPTION
**Pull request by fritzo**
Pair coded with @eb8680 

This factors out an `infer_discrete` poutine from the `collapse-messenger` branch. The future `collapse` messenger is a thin wrapper around `infer_discrete`; I will rebase that and the `mis_ssm` branches once this merges.

The new `infer_discrete` poutine is a stand-alone inference algorithm in models with only discrete latent variables (observed variables can have any type). Applications include Viterbi-like MAP estimation and forward-filter backward-sample inference. Eventually `TraceEnum_ELBO.smaple_posterior()` will use `infer_discrete`, reducing its complexity from quadratic to linear.

The algorithm is parametrized by a `temperature` variable that can be either temperature=1 for sampling from the posterior, or temperature=0 for MAP estimation.

## Tested

- added unit tests with a variety of model shapes

### CODE REVIEW

**Comment by martinjankowiak** [comment]
cool! should we use the name `posterior` for something rather specific?

**Comment by fritzo** [comment]
I welcome a better name suggestion :smile:

**Comment by fritzo** [comment]
How about one of the following:
- `infer_discrete`
- `transduce_discrete`
- `impute_discrete`
- `impute`
- `sample_posterior`

I think this might also better belong in `pyro.infer` than in `pyro.poutine`, since it is really a high-level inference algorithm built from smaller poutines...

File: tutorial/source/enumeration.ipynb [hunk 1]
```diff
@@ -207,6 +210,49 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
+    "### Examining discrete latent states <a class=\"anchor\" id=\"Examining-discrete-latent-states\"></a>\n",
+    "\n",
+    "While enumeration in SVI allows fast learning of parameters like `p` above, it does not give access to predicted values of the discrete latent variables like `x,y,z` above. We can access these using a standalone `infer_discrete` poutine. In this case the guide was trivial, so we can simply wrap the model in `infer_discrete`. We need to pass a `first_available_dim` argument to tell `infer_discrete` which dimensions are available for enumeration; this is related to the `max_plate_nesting` arg of `TraceEnum_ELBO` via\n",
```
Comment (eb8680) [inline tutorial/source/enumeration.ipynb]: Can you mention that score-function gradients through `infer_discrete` won't be correct? Also a nit: "standalone `infer_discrete` poutine" -> "standalone `infer_discrete` handler"

**Pull request merged**